starport/pkg/cosmosclient/cosmosclient.go

```diff
@@ -17,12 +21,169 @@ import (
 )
 
 const (
+	defaultNodeAddress   = "http://localhost:26657"
 	defaultGasAdjustment = 1.0
 	defaultGasLimit      = 300000
 )
 
-// NewContext creates a new client context.
-func NewContext(
+// KeyringBackend is the backend for where keys are stored.
+type KeyringBackend string
+
+const (
+	// KeyringTest is the test keyring backend. with this backend, your keys will be
+	// stored under your app's data dir,
+	KeyringTest KeyringBackend = "test"
+
+	// KeyringOS is the OS keyring backend. with this backend, your keys will be
+	// stored in your operating system's secured keyring.
+	KeyringOS KeyringBackend = "os"
+)
+
+// Client is a client to access your chain by querying and broadcasting transactions.
+type Client struct {
+	// RPC is Tendermint RPC.
+	RPC *rpchttp.HTTP
+
+	// Factory is a Cosmos SDK tx factory.
+	Factory tx.Factory
+
+	// context is a Cosmos SDK client context.
+	Context client.Context
+
+	// Keyring is a Cosmos SDK keyring.
+	Keyring keyring.Keyring
+
+	nodeAddress        string
+	out                io.Writer
+	chainID            string
+	homePath           string
+	keyringServiceName string
+	keyringBackend     KeyringBackend
+}
+
+// Option configures your client.
+type Option func(*Client)
+
+// WithNodeAddress sets the node address of your chain. when this option is not provided
+// `http://localhost:26657` used as default.
+func WithNodeAddress(addr string) Option {
+	return func(c *Client) {
+		c.nodeAddress = addr
+	}
+}
+
+// WithHome sets the data dir of your chain. this option is used to access to your chain's
+// file based keyring which is only needed when you deal with creating and signing transactions.
+// when it is not provided, your data dir will be assumed as `$HOME/.your-chain-id`.
+func WithHome(path string) Option {
+	return func(c *Client) {
+		c.homePath = path
+	}
+}
+
+// WithKeyringServiceName used as the keyring's name when you are using OS keyring backend.
+// by default it is `cosmos`.
+func WithKeyringServiceName(name string) Option {
+	return func(c *Client) {
+		c.keyringServiceName = name
+	}
+}
+
+// WithKeyringBackend sets your keyring backend. by default it is `test`.
+func WithKeyringBackend(backend KeyringBackend) Option {
+	return func(c *Client) {
+		c.keyringBackend = backend
+	}
+}
+
+// New creates a new client with given options.
+func New(ctx context.Context, options ...Option) (Client, error) {
+	c := Client{
+		nodeAddress:        defaultNodeAddress,
+		out:                io.Discard,
+		keyringServiceName: sdktypes.KeyringServiceName(),
+		keyringBackend:     KeyringTest,
+	}
+
+	var err error
+
+	for _, apply := range options {
+		apply(&c)
+	}
+
+	if c.RPC, err = rpchttp.New(c.nodeAddress, "/websocket"); err != nil {
+		return Client{}, err
+	}
+
+	statusResp, err := c.RPC.Status(ctx)
+	if err != nil {
+		return Client{}, err
+	}
+
+	c.chainID = statusResp.NodeInfo.Network
+
+	if c.homePath == "" {
+		c.homePath = os.ExpandEnv(fmt.Sprintf("$HOME/.%s", c.chainID))
```

Comment:
We should use `os.UserHomeDir()` to get home

Reply:
dec80a65